### PR TITLE
Design - 4598 - Alert component redesign

### LINKED ISSUE #4598: ⚙️ Alert component
# ⚙️ Summary

### [Visit the Figma file](https://www.figma.com/file/T1R64oNWog10OXiBnuiMkz/GC-Digital-Talent-Design-Standards?node-id=111%3A1334&t=oVTB4H7vWLVWuIm6-1)

A component that allows us to display several different types of messages to the user with levels of severity. All alerts should have a title and some form of body content, with the option of adding a footer (see below). Content should be succinct and to the point, but can contain interactive elements like links if necessary.

**For all layout and spacing guidance, including how this component should respond on smaller devices, please see the Figma file.**

## ✅ Design checklist
- [x] Default component
- [x] Dark mode component
- [x] Component variants
- [x] Dark mode component variants (if applicable)
- [x] Foundational design info (e.g. color info, icons, helpful values)
- [x] Mobile (`base` query) layout and space diagram
- [x] Desktop (`p-tablet` query or larger) layout and space diagram
- [x] Accessibility considerations
- [x] Hover states (if applicable)
- [x] Dark mode hover states (if applicable)
- [x] Focus states (if applicable)
- [x] Dark mode focus states (if applicable)

## ✅ Accessibility checklist
- [x] Tap targets are adequate on smaller screens
- [x] Color contrast for all modes and states has been checked and cleared
- [x] Zoom has been considered as layout changes occur
- [x] `alt` text has been identified and labelled
- [x] Non-standard states or interactions have been described and provided accessible design context

# 🛠️ Properties

## Alert type

image

**Info:** An alert styled in our `primary` color with a bell icon that allows the UI to communicate non-destructive, non-critical information that might be helpful to the context or task.

**Success:**  An alert styled in the `success` color with a checkmark icon that allows the UI to communicate that a task has successfully completed. For example a form has properly been submitted, or a setting has been properly saved.

**Warning:**  An alert styled in the `warning` color with a circular exclamation icon that allows the UI to warn against an action or behaviour. More often than not, warnings will be used as moderately-critical reminders that an action might be destructive.

**Error:**  An alert styled in the `error` color with a triangular exclamation icon that communicates errors in state or information. Most often, error alerts will be found when a form is submitted with invalid content.

## Persistent vs. dismissive

Alerts should be persistent by default. In specific circumstances however, it is helpful for the user to be able to close or dismiss the alert. By enabling a dismissive alert, an absolutely positioned button appears in the top-right corner that removes the alert from view. This is particularly helpful for transient information or success messages that don't provide time-sensitive context.

image

## Footer content

Footer content should be optional, and allows for extra contextual information or actionable side tasks such as contact information.

image

### Relative alerts

Relatively positioned alerts are the most common and exist within the page UI. Because toast alerts are often timed/transient, all information that requires an alert that is critical or actionable by the user should be placed in a static alert somewhere on the page. This applies to things like form validation, maintenance messages, etc.

image

### Toast alerts

Toast alerts are helpful for providing information that might not be related to information currently in view on the page. Because they are pinned to the viewport, they follow the user as they scroll, meaning that they can travel to the problem interface and bring the information with them.

Toast alerts shouldn't be transient by default, but should have the option to be disappear after a certain amount of time has passed if the context demands it.

They should be positioned and fixed to the lower right corner of the viewport to avoid blocking critical navigation elements and page information. When multiple toasts are activated, they should be stacked in order of arrival, from bottom to top. Dismissing an alert in the stack should bring the remaining alerts down to fill the space until they are are all removed.

#### As alerts arrive

image

#### As alerts leave

image

# 💡 Interaction states

The alert component only contains one unique element that changes based on state: the dismiss button child component.

## Hover

When hovered, the dismiss button child component should stay in place, but a larger background should be highlighted in the matching light color that corresponds to the alert type property (e.g. light red for error, light green for success, etc.). In the same vein, the icon itself should transition to the matching dark color.

image

## Focus

When focused, the dismiss button child component should stay in place, but its larger background should be highlighted in the system focus color, while the icon itself should transition to black to ensure a high contrast.

image

# 🥰 Accessibility

## Required `alt` text

- Icons for each alert type should have `alt` text containing the alert's type.
- If the alert is dismissive, the dismiss button child component should also have accessible labelling that describes the action.

## Accessible toast alerts

When a toast alert arrives on screen, assistive technologies should be notified as an interruption and provided the ability to focus the alert. Leaving the alert's focus should return the user to where they were, unless the alert itself has provided a focusable location on the page as a part of its context.

### PR DESCRIPTION
**Pull request by esizer**
## 👋 Introduction

This replaces the existing `<Alert />` component with the new design and adds `jest-axe` tests.

## 🧪 Testing

- Open chromatic and confirm diff
- Test dismissible
- Confirm all existing alerts updated with new API

## 📷 Screenshot

Screenshot 2022-11-08 at 5 01 06 PM

## 🤖 Robot Stuff

Resolves #4598

### CODE REVIEW

**Comment by petertgiles** [comment]
LOL, our hero returned!
image

File: frontend/common/src/components/Alert/Alert.tsx [hunk 1]
```diff
@@ -1,52 +1,131 @@
...
+type AlertHeadingLevel = HeadingLevel | "p";
 
-export interface AlertProps {
-  title: React.ReactNode | string;
-  children: React.ReactNode;
-  icon: React.FC<React.SVGProps<SVGSVGElement>>;
-  type: Color;
-  mode?: AlertMode;
+export interface AlertProps
+  extends Omit<React.HTMLProps<HTMLDivElement>, "title"> {
```
Comment (petertgiles) [inline frontend/common/src/components/Alert/Alert.tsx:21]: Could you please add some comments on these props?  Most are pretty clear but I'm not sure what "live" does and not 100% clear on heading level either.

**Comment by esizer** [comment]
> Don't forget your translations.

Yeah, I should have noted in the description. I sent this off to the designers to review and test for a11y. Not sure if that will be the final text yet 😬

**Review by petertgiles — APPROVED** [review]
Nice, it looks great!  I left a comment below but it's not huge.  
Also, two things before merging:
- Don't forget the translations.  The CI will green up when done.
- I've asked on Slack for Maddy to review, too.  She's out today so hopefully she'll have a look tomorrow.

File: frontend/common/src/components/Alert/Alert.tsx (hunk 1 above)
Comment (esizer) [reply frontend/common/src/components/Alert/Alert.tsx:32]: Great idea! Documented with [document alert props interface](https://github.com/GCTC-NTGC/gc-digital-talent/pull/4668/commits/203c1b905cf1f0ee1b0b738290c38e9680ee489c)

Comment (petertgiles) [reply frontend/common/src/components/Alert/Alert.tsx:32]: :man_cook: :pinched_fingers:

**Comment by esizer** [comment]
Sorry Peter, I had some good conversations with the designers today which lead me to updating the API for this component so much it needs a re-review.

They wanted to ability to add a footer and instead of making it a manual process, this was updated to use context so we can style the sub-components without passing the type to each child.

I'll add some docs in the morning but it will look something like this now:

## New Syntax

```tsx
<Alert.Root type="info">
  <Alert.Title>Some title</Alert.Title> 

  <p>Anything can go here.</p>

  <Alert.Footer>
    <p>Some call to action.</p>
  </Alert.Footer>
</Alert.Root>
```

**Review by petertgiles — APPROVED** [review]
Man those alerts are chonky!  :pig2: I do like the new subcomponent structure.

**Pull request merged**